### PR TITLE
Create TaggedAdminInterface

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because BC.

## Changelog

```markdown
### Added
- Added `TaggedAdminInterface`
- Added `AbstractTaggedAdmin`
- Added `AbstractAdmin::hasFilterPersister()` method

### Deprecated
- Deprecate instantiating a new `AbstractAdmin` class with `null` as third argument
- Deprecate calling `AbstractAdmin::getManagerType()` when no manager type is set
- Deprecate calling `AbstractAdmin::getFilterPersister()` when no filter persister is set
- Deprecate calling `AbstractAdmin::getModelManager()` when no model manager is set
- Deprecate calling `AbstractAdmin::getDataSource()` when no data source is set
- Deprecate calling `AbstractAdmin::getFormContractor()` when no form contractor is set
- Deprecate calling `AbstractAdmin::getShowBuilder()` when no show builder is set
- Deprecate calling `AbstractAdmin::getListBuilder()` when no list builder is set
- Deprecate calling `AbstractAdmin::getDatagridBuilder()` when no datagrid builder is set
- Deprecate calling `AbstractAdmin::getTranslator()` when no translator is set
- Deprecate calling `AbstractAdmin::getConfigurationPool()` when no pool is set
- Deprecate `AbstractAdmin::setValidator()`
- Deprecate `AbstractAdmin::getValidator()`
- Deprecate calling `AbstractAdmin::getSecurityHandler()` when no security handler is set
- Deprecate calling `AbstractAdmin::getMenuFactory()` when no menu factory is set
- Deprecate calling `AbstractAdmin::getRouteBuilder()` when no route builder is set
- Deprecate calling `AbstractAdmin::getLabelTranslatorStrategy()` when no label translator strategy is set
```

## To do (in another PR)

- Take a look at SideMenu/TabMenu

### CODE REVIEW

File: src/Admin/AdminTagInterface.php
Comment (core23) [inline src/Admin/AdminTagInterface.php:41]: This interface is still very big. I would prefer some more smaller interface for different use cases: render edit form(s), show list, exporter, add routes, ...

File: src/Admin/AdminTagInterface.php
Comment (core23) [inline src/Admin/AdminTagInterface.php:37]: Do we really need this method?

File: src/Admin/AdminTagInterface.php
Comment (core23) [inline src/Admin/AdminTagInterface.php:36]: An interface should not have a `__construct` method

**Review by core23 — CHANGES_REQUESTED** [review]

File: src/Admin/AdminTagInterface.php
Comment (VincentLanglet) [reply src/Admin/AdminTagInterface.php:36]: I understand, but I don't really know what to do.

Fact is the first and third parameter of the constructor are set by the AddDependencyCallsCompilerPass.
So if you don't have code as the first argument and controller as the third ; it won't work.

File: src/Admin/AdminTagInterface.php
Comment (VincentLanglet) [reply src/Admin/AdminTagInterface.php:37]: It's called by the AddDependencyCallsCompilerPass, so it's needed

File: src/Admin/AdminTagInterface.php
Comment (VincentLanglet) [reply src/Admin/AdminTagInterface.php:41]: As soon as you use the `sonata.admin` tag, you need all these methods since they are all called in the AddDependencyCallsCompilerPass.

**Comment by VincentLanglet** [comment]
@phansys We had a discussion about AddDependencyCallsCompilerPass.

After looking at this class, I discovered a lot of setter are needed by the AdminInterface if the class use the `sonata.admin` tag. So I created an AdminTagInterface with all the getter/setter need for this compiler pass.
I'm open to any improvement (especially for the name). I also created an AbstractAdminTag in order to split the AbstractAdmin.

This PR is ready to be reviewed @sonata-project/contributors I'm open to everything.

**Comment by phansys** [comment]
Since the tagging concept is only related to Symfony's DI, I think `AbstractAdmin` should not extend `AbstractAdminTag`. In this case, I'd expect `AbstractAdminTag` to extend or decorate `AbstractAdmin`.
That said, maybe we should use a name agnostic to the tagging concept.

**Comment by phansys** [comment]
>This interface is still very big. I would prefer some more smaller interface for different use cases: render edit form(s), show list, exporter, add routes, ...

I agree with this premise.
Maybe we could provide extensions or traits for these "extra" responsibilities and let `AbstractAdmin` to implement all of those `*AwareInterface` interfaces.

**Comment by VincentLanglet** [comment]
> Since the tagging concept is only related to Symfony's DI, I think `AbstractAdmin` should not extend `AbstractAdminTag`.

I think all this interface and all this abstract class are doing like people are usually creating their own AdminInterface class, but nobody does. A lot of methods were missing in the interface and there was no issue reporting this.

But without this,
- The `AddDependencyCallsCompilerPass` need alls the setter in the `AdminTagInterface`.
- The `AbstractAdminTag` provides a basic implementation of the `AdminTagInterface`.
- The `AbstractAdmin` is coded like all the getter are not returning `null` ; so it basically expects that the class is tag as a `sonata.admin`. Even if, you don't use the tag, you HAVE TO set all the dependencies manually to correctly use the class.

Maybe another name could satisfy you, but to me the `AbstractAdminTag/AdminTagInterface` is "All the dependencies you need to inject manually or not" to use the AdminInterface. 

> That said, maybe we should use a name agnostic to the tagging concept.

Maybe then, any suggestion now I gave you more explanation about how I see this ?

> > This interface is still very big. I would prefer some more smaller interface for different use cases: render edit form(s), show list, exporter, add routes, ...
> 
> I agree with this premise.
> Maybe we could provide extensions or traits for these "extra" responsibilities and let `AbstractAdmin` to implement all of those `*AwareInterface` interfaces.

I listed every responsibilities provided by the `AddDependencyCallsCompilerPass`. Splitting the Interface, when the compiler is not, does not make sens to me.

Currently there is just a huge AbstractAdmin class and a huge AdminInterface.
I split them in two ; that's still better than no split at all. 
We can do better in the next PR for sure ; but it already took me a Week end for cleaning this.

**Comment by franmomu** [comment]
Another possibility could be maybe building a custom `ServiceLocator` with the needed services for each Admin in the compiler pass and inject it to each Admin service. 

That way we could probably remove all setter and getters (if they are not used).

I haven't given much thought, but this is one of the things I wanted to try with services locators.

**Comment by VincentLanglet** [comment]
> Another possibility could be maybe building a custom `ServiceLocator` with the needed services for each Admin in the compiler pass and inject it to each Admin service. 
> 
> 
> 
> That way we could probably remove all setter and getters (if they are not used).
> 
> 
> 
> I haven't given much thought, but this is one of the things I wanted to try with services locators.

Not sure it's made for this use case.

Some setters are used in order to override a service ; for instance the AdminExtractor is overriding the TranslatorLabelTrategy. 

And a lot of getters are used in others services.

Here, I see the AdminTagInterface like the "configuration" of the admin. We should allow to override it or defining it manually.

**Comment by VincentLanglet** [comment]
I updated the Changelog.

I trying to get things cleaner for the Sonata 4 version. Currently
- The setter were needed for the `sonata.admin` tag but were missing in the Interface.
- The construct need his first argument to be the `code`  and the third to be the `controller` since they are injected.
- The getter are used like they never return `null`.
- The translator property is deprecated without any solution for the translation of the export labels.
- The admin property are typehinted as non-nullable when most of them are.

Note that some of the getter/setter are already added in the master branch.
And some of the getter are already throwing an exception if the value is null in the master branch.
This PR just do the same for all the required/injected dependencies.
After introducing the deprecation in 3.x, I'll do the merge in master and resolve the conflict.

This PR seems to me a clean (at least cleaner than now), BC and minimal way to fix all these 5 issues.

If for Sonata 5 we're refactoring everything in order to use a service locator or something else, it will easy to deprecate this.

I really would like to release Sonata 4 soon and I'm trying to clean some typehint/interface before.
See https://github.com/sonata-project/SonataAdminBundle/pull/6541 where a lot of methods are missing in the interface too.

**Comment by wbloszyk** [comment]
`AbstractAdmin` consists of:
- entity panel configuration
- list module
- show module
- edit module
- etc. 

Mybe it is possible to clean up (and split `AbstractAdmin`) by this parts? 

Technicly in feature we can split `SonataAdminBundle` into:
- `SonataCRUDBundle` - tools to create CRUD in front page (without global layout, people can create list with serach - ListModule... etc.)
- `SonataAdminBundle`

**Comment by VincentLanglet** [comment]
> `AbstractAdmin` consists of:
> 
> * entity panel configuration
> * list module
> * show module
> * edit module
> * etc.
> 
> Maybe it is possible to clean up (and split `AbstractAdmin`) by this parts?
> 
> Technically in feature we can split `SonataAdminBundle` into:
> 
> * `SonataCRUDBundle` - tools to create CRUD in front page (without global layout, people can create list with serach - ListModule... etc.)
> * `SonataAdminBundle`

I agree there is lot of way possible to split the AbstractAdmin, but it means
- Lot of work
- Lot of BC-break
There is already some issue like this about it https://github.com/sonata-project/SonataAdminBundle/issues/3947, and they have for milestone the 5.0 version, when we didn't even release the 4.0.

In 3.x, only the `setTemplate/getTemplate` method were deprecated, and it lead to lot of work (https://github.com/sonata-project/SonataAdminBundle/pull/6566) which is currently not finish and is one of the blocking point for the next major release. So I don't think it's a good idea to try to do other similar changes.

As said, here I tried to provide the quickest, simplest proper way to clean the situation with as few BC as possible.

I would like more some review on 
- The name of the new interface. Maybe it should be `AdminConfigurationInterface` ? Something else ?
- The AbstractAdminTag class. I never used trait, but maybe should it be a Trait ?
- The current issue with the `constructor` and the `AddDependencyCallsCompilerPass`. The CompilerPass is injecting automatically the `first` and the `third` argument, so to use the `sonata.admin` tag, a class should have a specific construct signature. But adding the constructor to the interface is not a good thing ; so maybe there is something to change in the compiler pass ?

**Comment by wbloszyk** [comment]
I can remove `AddDependencyCallsCompilerPass::fixTemplates()` after #6566 (Now I working on it, functionality is done, I must remove self deprecation usages).

**Comment by VincentLanglet** [comment]
> * The current issue with the `constructor` and the `AddDependencyCallsCompilerPass`. The CompilerPass is injecting automatically the `first` and the `third` argument, so to use the `sonata.admin` tag, a class should have a specific construct signature. But adding the constructor to the interface is not a good thing ; so maybe there is something to change in the compiler pass ?

For instance, 
```
public __construct(EntityManager $em, $code, $class, $baseControllerName)
```
Is not a valid construct signature with the sonata.admin tag. But
```
public __construct($code, $class, $baseControllerName, EntityManager $em)
```
Seems to be valid.

**Comment by VincentLanglet** [comment]
I don't have better way to solve the construct issue than https://github.com/sonata-project/SonataAdminBundle/pull/6614/commits/ef3b54ea6398ad6475f9fbb41148a7c45bfb8806

**Comment by VincentLanglet** [comment]
I removed the construct from the interface but added a note in the comments instead.

**Comment by phansys** [comment]
If we use the term "tag" in this interface, I think it must be declared under `Sonata\AdminBundle\DependencyInjection\` namespace, since it is absolutely related to the DI scope https://github.com/sonata-project/SonataAdminBundle/pull/6614#issuecomment-732133979.

**Review by phansys — APPROVED** [review]
I propose `AdminTaggedInterface`, but not a blocker at all.

**Comment by OskarStark** [comment]
What about TaggedAdminInterface?

Or can we rename it to me more speakable at all, what does it do? It what can you do after implementing this interface in one sentence?

**Comment by VincentLanglet** [comment]
@phansys @OskarStark It's renamed

**Review by OskarStark — APPROVED** [review]

File: src/Admin/AbstractAdmin.php
Comment (core23) [inline src/Admin/AbstractAdmin.php:56]: This looks more like a trait to me. We should not overuse class extension

**Review by core23 — CHANGES_REQUESTED** [review]

File: src/Admin/AbstractAdmin.php
Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php:56]: I dont think so.

A Trait can't have a construct @core23

And a class with the sonata.admin MUST have the following signature for the first constructor arguments.

Comment (core23) [reply src/Admin/AbstractAdmin.php:56]: 👍 That's a good indicator for a design issue (that we have for a long time...). If we need a specific signature for the constructor, we should not use it and create a new `initSomething` method for this.

Comment (core23) [reply src/Admin/AbstractAdmin.php:56]: Maybe we can use the `initialize` method for this

Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php:56]: > 👍 That's a good indicator for a design issue (that we have for a long time...). If we need a specific signature for the constructor, we should not use it and create a new `initSomething` method for this.

When defining an admin, we're writing 
```
    admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [~, App\Entity\Category, ~]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
The `arguments` key, is by definition the arguments of the `__construct()` method,
That's how service declaration is working for symfony, and that's how it's currently working for Sonata.

If it's a design issue, it's here from the first day. I introduced nothing.
You're just discovering or re-discovering this potential design issue, because of the clean/refacto.
So it shouldn't be a blocking point of this PR. Do you agree @core23 ?

Moreover I'm just trying to cleaning interface before Sonata 4.
The last refacto started is about `TemplateRegistry` and since it's not finished it's blocking the major release.
It's not time for another big refacto IMHO, and should be waiting Sonata 5 then.

Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php:56]: > Maybe we can use the `initialize` method for this

IMHO, we could just migrate from
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [~, App\Entity\Category, ~]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
to
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [App\Entity\Category]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```

And from 
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [CustomCode, App\Entity\Category, CustomController::class]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
to 
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [App\Entity\Category]
        tags:
            - { name: sonata.admin, code: CustomCode, controller:  CustomController::class, manager_type: orm, label: Category }
```

WDYT ?

**Review by OskarStark — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
Is it ok for you @core23 ?

File: src/Admin/AbstractAdmin.php
Comment (phansys) [reply src/Admin/AbstractAdmin.php:56]: Seems like a good solution to me :+1:

Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php:56]: @core23 Is it ok for you to delay this in another PR/milestone ?

The design issue is here for a long time like you said and shouldn't be a blocker here. I can't solve everything.
We just have to do small steps. Especially when we wants to release 4.0.

**Comment by VincentLanglet** [comment]
@sonata-project/contributors Can I get some approval ? :)

**Review by OskarStark — APPROVED** [review]
👌🏻 from my side

**Review by phansys — APPROVED** [review]
Please, update the changelog messsages.
By instance: `AbstractAdmin:: getManagerType` => `AbstractAdmin::getManagerType()`.

**Comment by VincentLanglet** [comment]
I updated the changelog @phansys

@greg0ire @wbloszyk I have no answer for 10 days from core23, his first request is done, the second one is not doable. Since I have 2 approval, would it be ok to dismiss the request changes ?

**Comment by phansys** [comment]
>I updated the changelog @phansys

I think the parentheses are missing in the methods' references.

**Comment by OskarStark** [comment]
Do you prefer a Squash or rebase merge @VincentLanglet ?

**Comment by VincentLanglet** [comment]
> Do you prefer a Squash or rebase merge @VincentLanglet ?

I think the rebase could help to understand some change, but I'm ok with a squash if you prefer @OskarStark

**Comment by OskarStark** [comment]
Rebase is fine 👍 

Thank you @VincentLanglet !

**Pull request merged**
